auth_totp/models/res_users_authenticator.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2016-2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import logging
+from odoo import _, api, fields, models
+
+_logger = logging.getLogger(__name__)
+try:
+    import pyotp
+except ImportError:
+    _logger.debug(
+        'Could not import PyOTP. Please make sure this library is available in'
+        ' your environment.'
+    )
+
+
+class ResUsersAuthenticator(models.Model):
+    _name = 'res.users.authenticator'
+    _description = 'MFA App/Device'
+    _sql_constraints = [(
+        'user_id_name_uniq',
+        'UNIQUE(user_id, name)',
+        _(
+            'There is already an MFA app/device with this name associated with'
+            ' your account. Please pick a new name and try again.'
+        ),
+    )]
+
+    name = fields.Char(
+        required=True,
+        readonly=True,
+    )
+    secret_key = fields.Char(
+        required=True,
+        readonly=True,
+    )
+    user_id = fields.Many2one(
+        comodel_name='res.users',
+        ondelete='cascade',
+    )
+
+    @api.multi
+    @api.constrains('user_id')
+    def _check_has_user(self):
+        self.filtered(lambda r: not r.user_id).unlink()
```

Comment:
a constrain that deletes?

Reply:
That's from previous version. What should I do?

Reply:
I dunno... :) How is it supposed to work? Who's resetting user_id?

Reply:
@obulkin Perhaps you know better. But to me, user_id should be a required field in 'res.users.authenticator'. Then this @api.constrains would not be required at all.

Reply:
The authenticators are only exposed through an o2m field in a user view, which doesn't support deletion, so this logic is here to clean up orphaned authenticators when they're unmapped from a user via the view. You could make the field required and drop the `constrains`, but you would need to change the way that users manage their authenticators through the UI.